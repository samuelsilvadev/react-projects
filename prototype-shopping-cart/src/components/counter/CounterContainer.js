import React from 'react';
import Counter from './Counter';

import './Counter.css';

const mockCounters = [
	{
		id: 1,
		value: 2,
	},
	{
		id: 2,
	},
	{
		id: 3,
	}
];

class CounterContainer extends React.Component {

	state = {
		counters: mockCounters,
	}

	render() {
		const { counters } = this.state;
		return (
			<React.Fragment>
				<button
					data-test="counter-container-btn-reset"
					className="btn btn-primary btn-sm btn-reset"
					disabled={!counters.length}
					onClick={this._handleClickResetBtn}>
					Reset
				</button>
				{counters.map(this._renderCounter, this)}
			</React.Fragment>
		);
	}

	_renderCounter(data) {
		const { id, value } = data;
		const { _removeCounter, _handleClickIncrementBtn } = this;
		return <Counter
			key={id}
			id={id}
			value={value}
			onIncrement={_handleClickIncrementBtn}
			onDelete={_removeCounter} />;
	}

	_handleClickResetBtn = () => {
		const { _resetAllCounters } = this;
		_resetAllCounters();
	}

	_handleClickIncrementBtn = (id) => {
		const { counters } = this.state;
		const newCounters = counters.map(counter => {
			if (counter.id === id) {
				if (counter.value) {
					counter.value++;
				} else {
					counter.value = 1;
				}
			}
			return counter;
		});

		this.setState({
			counters: newCounters,
		});
	}

	_removeCounter = (id) => {
		const { counters } = this.state;
		const newCounters = counters.filter(counter => counter.id !== id);

		this.setState({
			counters: newCounters,
		});
	}

	_resetAllCounters = () => {
		const { counters } = this.state;
		const newCounters = counters.map(counter => {
			counter.value = 0;
			return counter;
		});

		this.setState({
			counters: newCounters,
		});
	}
};

export default CounterContainer;
