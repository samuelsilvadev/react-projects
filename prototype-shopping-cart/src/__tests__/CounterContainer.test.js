import React from 'react';
import { shallow } from 'enzyme';

import CounterContainer from '../components/counter/CounterContainer';

describe('<CounterContainer />', () => {
    it('should render correctly without props', () => {
        const CounterContainerComponent = shallow(<CounterContainer />);

        expect(CounterContainerComponent).toMatchSnapshot();
	});
	
	it('should render correctly with props', () => {
		const counters = [
			{
				id: 1,
				value: 2,
			},
			{
				id: 2,
				value: 0,
			},
			{
				id: 3,
			}
		];
		const CounterContainerComponent = shallow(<CounterContainer />);
		CounterContainerComponent.setState({ counters });
		const instance = CounterContainerComponent.instance();

		expect(instance.state.counters.length).toBe(counters.length);
        expect(CounterContainerComponent).toMatchSnapshot();
    });
});
