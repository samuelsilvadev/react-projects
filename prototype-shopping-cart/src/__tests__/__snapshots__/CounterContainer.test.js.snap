// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CounterContainer /> should render correctly with props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CounterContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <button
          className="btn btn-primary btn-sm"
          onClick={[Function]}
        >
          Reset
        </button>,
        Array [
          <Counter
            id={1}
            onDelete={[Function]}
            onIncrement={[Function]}
            value={2}
          />,
          <Counter
            id={2}
            onDelete={[Function]}
            onIncrement={[Function]}
            value={0}
          />,
          <Counter
            id={3}
            onDelete={[Function]}
            onIncrement={[Function]}
            value={0}
          />,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Reset",
          "className": "btn btn-primary btn-sm",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Reset",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "id": 1,
          "onDelete": [Function],
          "onIncrement": [Function],
          "value": 2,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "id": 2,
          "onDelete": [Function],
          "onIncrement": [Function],
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "class",
        "props": Object {
          "id": 3,
          "onDelete": [Function],
          "onIncrement": [Function],
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <button
            className="btn btn-primary btn-sm"
            onClick={[Function]}
          >
            Reset
          </button>,
          Array [
            <Counter
              id={1}
              onDelete={[Function]}
              onIncrement={[Function]}
              value={2}
            />,
            <Counter
              id={2}
              onDelete={[Function]}
              onIncrement={[Function]}
              value={0}
            />,
            <Counter
              id={3}
              onDelete={[Function]}
              onIncrement={[Function]}
              value={0}
            />,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Reset",
            "className": "btn btn-primary btn-sm",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Reset",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "id": 1,
            "onDelete": [Function],
            "onIncrement": [Function],
            "value": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "id": 2,
            "onDelete": [Function],
            "onIncrement": [Function],
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "id": 3,
            "onDelete": [Function],
            "onIncrement": [Function],
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<CounterContainer /> should render correctly without props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CounterContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <button
          className="btn btn-primary btn-sm"
          onClick={[Function]}
        >
          Reset
        </button>,
        Array [
          <Counter
            id={1}
            onDelete={[Function]}
            onIncrement={[Function]}
            value={2}
          />,
          <Counter
            id={2}
            onDelete={[Function]}
            onIncrement={[Function]}
            value={0}
          />,
          <Counter
            id={3}
            onDelete={[Function]}
            onIncrement={[Function]}
            value={0}
          />,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Reset",
          "className": "btn btn-primary btn-sm",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Reset",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "id": 1,
          "onDelete": [Function],
          "onIncrement": [Function],
          "value": 2,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "id": 2,
          "onDelete": [Function],
          "onIncrement": [Function],
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "class",
        "props": Object {
          "id": 3,
          "onDelete": [Function],
          "onIncrement": [Function],
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <button
            className="btn btn-primary btn-sm"
            onClick={[Function]}
          >
            Reset
          </button>,
          Array [
            <Counter
              id={1}
              onDelete={[Function]}
              onIncrement={[Function]}
              value={2}
            />,
            <Counter
              id={2}
              onDelete={[Function]}
              onIncrement={[Function]}
              value={0}
            />,
            <Counter
              id={3}
              onDelete={[Function]}
              onIncrement={[Function]}
              value={0}
            />,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Reset",
            "className": "btn btn-primary btn-sm",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Reset",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "id": 1,
            "onDelete": [Function],
            "onIncrement": [Function],
            "value": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "id": 2,
            "onDelete": [Function],
            "onIncrement": [Function],
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "id": 3,
            "onDelete": [Function],
            "onIncrement": [Function],
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
